Attribute Systems: Critical, Thorns and Life Leech

Rev219 opened the pull request:
Full critical system implementation, should work on both entities and players.

immortius commented on src/main/java/com/wanderersoftherift/wotr/core/OnHitEffect.java:
```diff
@@ -0,0 +1,23 @@
+        int critAvoid = (int) foe.getAttributeValue(WotrAttributes.CRITICAL_AVOIDANCE);
+
+        int critChance = baseCritChance - critAvoid;
+        int critApplications = critChance / 100;
```
What happens if critChance is < -100?

immortius commented on src/main/java/com/wanderersoftherift/wotr/init/WotrAttributes.java:
```diff
@@ -81,9 +97,11 @@ public static List<DeferredHolder<Attribute, RangedAttribute>> getPlayerAttribut
     @SubscribeEvent
-    private static void addAttributesToPlayer(EntityAttributeModificationEvent event) {
+    private static void addAttributesToLivingEntities(EntityAttributeModificationEvent event) {
         for (DeferredHolder<Attribute, RangedAttribute> attribute : PLAYER_ATTRIBUTES) {
```
Probably should rename `PLAYER_ATTRIBUTES`, `registerPlayerAttribute`, and `getPlayerAttributes` for clarity.

Rev219 replied:
done

getPlayerAttributes is unused, and renamed both PLAYER_ATTRIBUTES and the aforementioned method

immortius commented on src/main/java/com/wanderersoftherift/wotr/core/OnHitEvent.java:
```diff
@@ -0,0 +1,18 @@
+
+    @SubscribeEvent
+    public static void onHit(LivingIncomingDamageEvent event) {
+        float dmg = OnHitEffect.critical(event.getAmount(), (LivingEntity) event.getSource().getDirectEntity(),
```
A technique I like to use is 
```java
if (event.getSource().getDirectEntity() instanceof LivingEntity livingAttacker) {
  // proceed using livingAttacker.   
}
```
This combines checking the directEntity is not null and is the right type with casting it.

immortius commented on src/main/java/com/wanderersoftherift/wotr/core/OnHitEvent.java:
```diff
@@ -10,9 +11,28 @@
+    }
+
+    @SubscribeEvent
+    public static void onThornsProc(LivingIncomingDamageEvent event) {
```
Would suggest method name would be better as `checkThornsProc` (it isn't called when thorns are triggered)

immortius commented on src/main/java/com/wanderersoftherift/wotr/core/OnHitEvent.java:
```diff
@@ -10,9 +11,28 @@
+    @SubscribeEvent
+    public static void onThornsProc(LivingIncomingDamageEvent event) {
+        if (event.getEntity() instanceof LivingEntity living) {
+            if (living.getLastAttacker() instanceof LivingEntity attacker) {
```
If the player is hit by a mob, and then 10 seconds later stands in a fire, should the mob be taking thorns damage from the fire hurting the player?

immortius commented on src/main/java/com/wanderersoftherift/wotr/core/OnHitEvent.java:
```diff
@@ -10,9 +11,28 @@
+        if (event.getEntity() instanceof LivingEntity living) {
+            if (living.getLastAttacker() instanceof LivingEntity attacker) {
+                boolean t = OnHitEffect.thorns(living, living.getRandom());
+                double thornsDamage = living.getAttributeValue(WotrAttributes.THORNS_DAMAGE);
```
Just noting there probably needs to be a check somewhere for whether the the causing damage is thorns, so that thorns doesn't trigger off thorns (resulting in either the game crashing or both entities exploding).

ImplementsLegend commented on src/main/java/com/wanderersoftherift/wotr/init/WotrAttributes.java:
```diff
@@ -81,9 +105,11 @@ public static List<DeferredHolder<Attribute, RangedAttribute>> getPlayerAttribut
-        for (DeferredHolder<Attribute, RangedAttribute> attribute : PLAYER_ATTRIBUTES) {
-            event.add(EntityType.PLAYER, attribute);
+    private static void registerMobAttribute(EntityAttributeModificationEvent event) {
+        for (DeferredHolder<Attribute, RangedAttribute> attribute : MOB_ATTRIBUTES) {
```
You are iterating over `MOB_ATTRIBUTES` but all attributes are in `PLAYER_ATTRIBUTES`. As a result no attribute gets added to any entity.

ImplementsLegend commented on src/main/java/com/wanderersoftherift/wotr/core/OnHitEvent.java:
```diff
@@ -0,0 +1,63 @@
+
+        Entity causer = event.getSource().getEntity();
+        LivingEntity reciever = event.getEntity();
+        if (!WotrDamageTypes.THORNS_DAMAGE.equals(event.getSource().typeHolder()) && causer != null) {
```
You should also check if `event.getAmount() == 0`.

immortius replied:
The equals check always returns false (left side is a ResourceKey, right side is a Holder).

ImplementsLegend replied:
Right, it should be:
```suggestion
        if (!event.getSource().typeHolder().is(WotrDamageTypes.THORNS_DAMAGE) && causer != null) {
```

ImplementsLegend replied:
Also thorns loop might still be possible if one entity has WotR thorns and the other has Minecraft thorns.

ImplementsLegend commented on src/main/java/com/wanderersoftherift/wotr/core/OnHitEvent.java:
```diff
@@ -10,9 +11,28 @@
-                event.getEntity(), event.getEntity().getRandom());
-        event.setAmount(dmg);
+    public static void onCriticalHit(LivingIncomingDamageEvent event) {
+        if (event.getSource().getDirectEntity() instanceof LivingEntity livingAttacker) {
```
I literally meant just:
```suggestion
        if (event.getSource().getEntity() instanceof LivingEntity livingAttacker) {
```

immortius replied:
I was checking how I got the correct entity to use wrong, and somewhere along the call chain in the Mojang side the causingEntity and directEntity swap variable names :/

immortius commented on src/main/java/com/wanderersoftherift/wotr/core/OnHitEffect.java:
```diff
@@ -0,0 +1,51 @@
+import net.minecraft.util.RandomSource;
+import net.minecraft.world.entity.LivingEntity;
+
+public class OnHitEffect {
```
Mmm... I would either just combine this class with OnHitEvent, or combine and then split out as three classes, one for each effect. None of this code is being reused outside of OnHitEvent, and there doesn't seem to be any driver for this code to be separated from the rest of the event handling.

Rev219 replied:
moved all 3 effects to a different class for each. they're located under effects now

immortius commented on src/main/java/com/wanderersoftherift/wotr/init/WotrAttributes.java:
```diff
@@ -24,6 +24,8 @@ public class WotrAttributes {
 
     private static final List<DeferredHolder<Attribute, RangedAttribute>> PLAYER_ATTRIBUTES = new ArrayList<>();
 
+    private static final List<DeferredHolder<Attribute, RangedAttribute>> MOB_ATTRIBUTES = new ArrayList<>();
```
Not used so could be removed

immortius commented on src/main/java/com/wanderersoftherift/wotr/init/WotrAttributes.java:
```diff
@@ -24,6 +24,8 @@ public class WotrAttributes {
 
     private static final List<DeferredHolder<Attribute, RangedAttribute>> PLAYER_ATTRIBUTES = new ArrayList<>();
```
With the change in usage would suggest renaming to just ATTRIBUTES (along with its usages. (shift + F6 in IntelliJ)

immortius commented on src/main/java/com/wanderersoftherift/wotr/init/WotrAttributes.java:
```diff
@@ -73,19 +75,41 @@ public class WotrAttributes {
-        }
-    }
 
     private static DeferredHolder<Attribute, RangedAttribute> registerPlayerAttribute(
```
I would rename this just to `registerAttribute` to reflect its changed behavior.

ImplementsLegend commented on src/main/java/com/wanderersoftherift/wotr/init/WotrAttributes.java:
```diff
@@ -73,19 +75,41 @@ public class WotrAttributes {
+            "life_leech",
+            () -> new RangedAttribute(WanderersOfTheRift.translationId("attribute", "life_leech"), 0, 0, 1F));
+
     public static List<DeferredHolder<Attribute, RangedAttribute>> getPlayerAttributes() {
```
On the topic of renaming, rename this too to `getAttributes`.

ImplementsLegend commented on src/main/java/com/wanderersoftherift/wotr/core/OnHitEvent.java:
```diff
@@ -0,0 +1,63 @@
+                level.registryAccess().lookupOrThrow(Registries.DAMAGE_TYPE).getOrThrow(WotrDamageTypes.THORNS_DAMAGE));
+
+        Entity causer = event.getSource().getEntity();
+        LivingEntity reciever = event.getEntity();
```
Spelling mistake
```suggestion
        LivingEntity receiver = event.getEntity();
```

Rev219 replied:
always get messed up with the ie and ei in that word lol

Rev219 commented on src/main/java/com/wanderersoftherift/wotr/core/OnHitEvent.java:
```diff
@@ -0,0 +1,63 @@
+            float health = causer.getHealth();
+
+            if (receiver.getHealth() < event.getAmount()) {
+                causer.setHealth(health + receiver.getHealth());
```
well, since pat wanted the life leech attribute to be between 0 and 1, i was trying to make it a multiplier for the actual health that gets healed in the process

Rev219 commented:
it failed due to using a .* import...

ImplementsLegend commented on src/main/java/com/wanderersoftherift/wotr/effects/CriticalEffect.java:
```diff
@@ -0,0 +1,29 @@
+                - (int) foe.getAttributeValue(Attributes.CRITICAL_AVOIDANCE);
+        float critMult = (float) attacker.getAttributeValue(Attributes.CRITICAL_MULTIPLIER);
+
+        int critApplications = 0;
+
+        if (critChance > 100) {
+            critApplications = critChance / 100;
+        } else {
+            if (random.nextInt(0, 100) < (critChance % 100)) {
+                critApplications += 1;
+            }
+        }
```
Please, stop being fixated on those ifs

```suggestion
        int critApplications = (critChance + random.nextInt(100)) / 100;
```

ImplementsLegend commented on src/main/java/com/wanderersoftherift/wotr/effects/ThornsEffect.java:
```diff
@@ -0,0 +1,21 @@
+public class ThornsEffect {
+    public static int calcThornsMult(LivingEntity entity, RandomSource random) {
+        int thornsChance = (int) entity.getAttributeValue(Attributes.THORNS_CHANCE);
+        int thornsApplications = 0;
+
+        if (thornsChance > 100 || thornsChance == 100) {
+            thornsApplications = thornsChance / 100;
+        } else {
+            if (random.nextInt(0, 100) < (thornsChance % 100)) {
+                thornsApplications += 1;
+            }
+        }
```
Here too.
```suggestion
    int thornsApplications = (thornsChance + random.nextInt(100)) / 100;
```

ImplementsLegend commented on src/main/java/com/wanderersoftherift/wotr/effects/ThornsEffect.java:
```diff
@@ -0,0 +1,21 @@
+import net.minecraft.world.entity.LivingEntity;
+
+public class ThornsEffect {
+    public static int calcThornsMult(LivingEntity entity, RandomSource random) {
```
Again, why does this only calculate multiplier when the other effects calculate total damage/health?

ImplementsLegend commented on src/main/java/com/wanderersoftherift/wotr/effects/events/OnHitEvent.java:
```diff
@@ -0,0 +1,59 @@
+        Entity causer = event.getSource().getEntity();
+        LivingEntity receiver = event.getEntity();
+        if (!event.getSource().is(DamageTypes.THORNS) && !event.getSource().is(WotrDamageTypes.THORNS_DAMAGE)
+                && causer != null && event.getOriginalDamage() != 0) {
```
It's better to use `getNewDamage` since some other event could've changed it.
```suggestion
                && causer != null && event.getNewDamage() != 0) {
```

ImplementsLegend commented on src/main/java/com/wanderersoftherift/wotr/effects/events/OnHitEvent.java:
```diff
@@ -0,0 +1,59 @@
+        LivingEntity causer = (LivingEntity) event.getSource().getEntity();
+        LivingEntity receiver = event.getEntity();
+        if (causer != null) {
+            causer.setHealth(LifeLeechEffect.calcHeal(causer, receiver, event.getOriginalDamage()));
```
Same here
```suggestion
            causer.setHealth(LifeLeechEffect.calcHeal(causer, receiver, event.getNewDamage()));
```

ImplementsLegend commented on src/main/java/com/wanderersoftherift/wotr/init/Attributes.java:
```diff
@@ -18,80 +18,111 @@
 @EventBusSubscriber(modid = WanderersOfTheRift.MODID, bus = EventBusSubscriber.Bus.MOD)
 @SuppressWarnings("checkstyle:DeclarationOrder")
-public class WotrAttributes {
+public class Attributes {
```
I have absolutely no idea why did you rename this.

ImplementsLegend commented on src/main/java/com/wanderersoftherift/wotr/init/Attributes.java:
```diff
@@ -18,80 +18,111 @@
     public static final DeferredRegister<Attribute> ATTRIBUTES = DeferredRegister.create(Registries.ATTRIBUTE,
             WanderersOfTheRift.MODID);
 
     private static final List<DeferredHolder<Attribute, RangedAttribute>> PLAYER_ATTRIBUTES = new ArrayList<>();
```
This, on the other hand, should have been renamed.

ImplementsLegend commented on src/main/java/com/wanderersoftherift/wotr/effects/events/OnHitEvent.java:
```diff
@@ -0,0 +1,59 @@
+        }
+    }
+
+    @SubscribeEvent
+    public static void lifeLeechEvent(LivingDamageEvent.Pre event) {
```
Add event priority.
```suggestion
    @SubscribeEvent(priority = EventPriority.LOW)
    public static void lifeLeechEvent(LivingDamageEvent.Pre event) {
```

Rev219 commented:
ok, so i fixed everything that you mentioned, and added thorns to projectiles impacting the player/mob. 

sorry about using so many if statements, had no idea i could do the things you mentioned (the if statement hell of thorns and critical) to make it easier

ImplementsLegend commented on src/main/java/com/wanderersoftherift/wotr/effects/events/OnHitEvent.java:
```diff
@@ -0,0 +1,73 @@
+                    causer.hurtServer((ServerLevel) level, thorns,
+                            (float) receiver.getAttributeValue(WotrAttributes.THORNS_DAMAGE));
+                }
+            } else if (causer instanceof Projectile proj) {
```
`event.getSource().getDirectEntity()` returns the projectile.
`event.getSource().getEntity()` returns the shooter. Therefore if somehow causer is projectile now, it implies that `proj.getOwner()==null` so calling hurtServer on the owner will crash with NullPointersException

Rev219 replied:
unsure on how to get the shooter to not be null..

any advice?

ImplementsLegend replied:
Are you asking how to get arrow without owner? Just shoot it from a dispenser.

ImplementsLegend replied:
Or like were you asking how to get the shooter when `event.getSource().getEntity()` is projectile? The point I was making is that you can't because shooter is not an entity.

ImplementsLegend commented on src/main/java/com/wanderersoftherift/wotr/effects/events/OnHitEvent.java:
```diff
@@ -0,0 +1,73 @@
+
+    @SubscribeEvent(priority = EventPriority.LOW)
+    public static void lifeLeechEvent(LivingDamageEvent.Pre event) {
+        Entity causer = event.getSource().getDirectEntity();
```
No leech for archers?

ImplementsLegend commented on src/main/java/com/wanderersoftherift/wotr/effects/events/OnHitEvent.java:
```diff
@@ -0,0 +1,73 @@
+    @SubscribeEvent
+    public static void thornsEvent(LivingDamageEvent.Post event) {
+        Level level = event.getEntity().level();
+        DamageSource thorns = new DamageSource(
```
DamageSource creation could be done later. Also could be named better.

ImplementsLegend commented on src/main/java/com/wanderersoftherift/wotr/effects/CriticalEffect.java:
```diff
@@ -0,0 +1,21 @@
+        float critMult = (float) attacker.getAttributeValue(WotrAttributes.CRITICAL_MULTIPLIER);
+        int critApplications = (critChance + random.nextInt(100)) / 100;
+
+        return initialDamage * (critMult * critApplications);
```
Why did you undo the change that I made?
```suggestion
        return initialDamage * ((critMult - 1) * critApplications + 1);
```
Final damage for damage multiplier 1.5:
| applications | current version | my version|
| -- | -- | -- |
| 0 | 0× | 1× |
| 1 | 1.5× | 1.5× |
| 2 | 3× | 2× |

We don't want 0 damage with 0 applications.

Maybe multiplier should be replaced with "bonus" that would start at 50%,. That would be effectively same as my version.
```suggestion
        return initialDamage + initialDamage * critBonus * critApplications;
```

ImplementsLegend replied:
In the second variation, the `critBonus` is *extra* damage on top of the base damage. For to get `critMiltiplier=1.5`, the bonus would be `0.5`. In general `critBonus = critMultiplier - 1`. And the first variation is the same, just written differently.

ImplementsLegend commented on src/main/java/com/wanderersoftherift/wotr/effects/events/OnHitEvent.java:
```diff
@@ -0,0 +1,73 @@
+        Entity causer = event.getSource().getEntity();
+        if (causer != null) {
+            LivingEntity living = (LivingEntity) causer;
+            if (living.getAttributeValue(WotrAttributes.CRITICAL_CHANCE) > living.getRandom().nextInt(0, 100)) {
```
Why is here a chance roll?

ImplementsLegend reviewed (changes requested):
It's getting better, just not quite there yet.

Rev219 commented on src/main/java/com/wanderersoftherift/wotr/entity/projectile/SimpleEffectProjectile.java:
```diff
@@ -445,7 +444,8 @@ protected void doKnockback(LivingEntity entity, DamageSource damageSource) {
         if (d0 > 0.0) {
-            double d1 = Math.max(0.0, 1.0 - entity.getAttributeValue(Attributes.KNOCKBACK_RESISTANCE));
+            double d1 = Math.max(0.0, 1.0 - entity
+                    .getAttributeValue(net.minecraft.world.entity.ai.attributes.Attributes.KNOCKBACK_RESISTANCE));
```
huh, on my intellij it says that it's already the import

ImplementsLegend replied:
```suggestion
                    .getAttributeValue(Attributes.KNOCKBACK_RESISTANCE));
```

Rev219 commented:
ok so post push comment. i also added some checks for if there was any damage reduction, and use the new damage if that's the case (since it was giving me 0 since there was none). otherwise it should use the original damage as a failsafe

ImplementsLegend commented:
`getNewDamage()` is what minecraft uses when applying damage to entity. If it is 0, it is supposted to be 0 and no damage is applied.

Rev219 commented on src/main/java/com/wanderersoftherift/wotr/effects/events/OnHitEvent.java:
```diff
@@ -0,0 +1,59 @@
+        Level level = event.getEntity().level();
+        Entity causer = event.getSource().getEntity();
+        LivingEntity receiver = event.getEntity();
+        DamageSource thornsDamageSource = new DamageSource(
+                level.registryAccess().get(WotrDamageTypes.THORNS_DAMAGE).get(), receiver, receiver, causer.position());
+        if (causer instanceof LivingEntity livCauser && causer.isAlive()) {
+            if (!event.getSource().is(WotrDamageTypes.THORNS_DAMAGE) && !event.getSource().is(DamageTypes.THORNS)
+                    && event.getNewDamage() != 0) {
+                livCauser.hurtServer((ServerLevel) level, thornsDamageSource,
+                        (float) ThornsEffect.calcThornsDamage(livCauser, livCauser.getRandom()));
```
ok, so i swapped causer.position and livCauser with receiver, if that's what you meant. also moved creating the damage source to be inside the same block where the thorns damage happens

Rev219 replied:
swapped .position to livCauser

ImplementsLegend reviewed (approved):
Seems good.
